Migrate definitions to a custom resource

iamjohnnym opened the pull request:
Since definitions are no longer recommended, migrating them to a custom resource seemed like a decent thing to do.  While the tests will pass, it throws an deprecation error which is related to `minitest`.  This error exists even with `definitions`.

```
         - MiniTest::Chef::Handler
       Running handlers complete

       Deprecated features used!
         Resource chef_handler from a cookbook is overriding the resource from the client. Please upgrade your cookbook or remove the cookbook from your run_list before the next major release of Chef. at 1 location:
           - /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.5.33/lib/chef/log.rb:51:in `caller_location'
          See https://docs.chef.io/deprecations_map_collision.html for further details.
```

chr4 commented on resources/openssh_server.rb:
```diff
@@ -52,7 +56,7 @@ class Chef::Recipe
     source    source
     variables config: sshd_config
-    action    template_action
+    action    :create
```
Hardcoding `:create` here might break rare cases where the old definition was used with an `:delete` or `:remove` action. I don't think supporting something like `:delete` is necessary, but this might be something we should mention in the CHANGELOG.

iamjohnnym replied:
I was concerned that this would be a blocker for that specific reason.  I considered adding as a `parameter` which could be set within the recipe.  I wasn't sure what the best approach would be.

chr4 commented:
Thanks for taking this on! The reason why I've hesitated for so long to migrate this was, that major purpose of this cookbook was to use specific defaults for each supported distribution and only overriding exactly the provided settings. This allowed e.g. changing the listening port without tampering with the different paths to `scp`. 

As it's hopeless to continue using the definition, I'm definitly in for changing this into a resource, but so far didn't find a way to mimic the above behaviour. Seems like you found one!

Is there anything you can think of that needs to be adapted in the README?

chr4 commented:
I'm glad one of my open source repositories got some hacktoberfest love 👍 !

I've just tested this on one of my machines, and I found out that the change unfortunately doesn't work when using the definition from a wrapper cookbook:

```ruby
openssh_server node['sshd']['config_file'] do
  Port 22
  PasswordAuthentication 'no'
  UseDNS 'no'
  X11Forwarding 'no'
end
```

I suppose the resource needs to be adapted (resp. a provider added?) so this cookbook can be used without having to be configured by node attributes? I'd be fine with a e.g. `params` or `sshd_config` parameter, so it can be called like this (or something similar):

```ruby
openssh_server node['sshd']['config_file'] do
  sshd_config = {
    Port: 22,
    PasswordAuthentication: 'no',
    UseDNS: 'no',
    X11Forwarding: 'no',
  }
end
```

chr4 commented:
(I'm guilty of not having provided tests for definition usage...)

iamjohnnym commented:
Hmm, I can do this but what's the difference between using a custom `attributesche` file over declaring them when instantiating the resource?

tbf, I'm really not sure.  I've been picking things up as I move along so my knowledge might be lacking.

chr4 commented:
Chef has two basic ways of configuring things. 

1. Via node attributes (the JSON file): Quick way of doing things, but also very limited, as there's no dependence on other attributes possible, let alone any logic
2. With wrapper cookbooks. Basically, the only thing you put into your node attributes is a specific cookbook and this cookbook then takes care of the rest. Which is basically the only way I use and recommend Chef cookbooks, because attributes are very limited in bigger/ more complex situations.

iamjohnnym commented:
Alright.  I believe I have addressed the issues with declaring additional params for the resource.  I also updated the README to reflect the new requirements.  While I did not write tests for the changes, i did test them.  I didn't write them because my next move will be doing `inspec` for testing.

```ruby
# Configure service
openssh_server node['sshd']['config_file'] do
  name            node['sshd']['config_file']
  sshd_config(
    UseDNS:      'no',
    Match:       {
      'User fred' => { 'X11Forwarding' => 'no' },
      'User john' => {
        'ChrootDirectory' => '/srv',
        'ForceCommand' => 'internal-sftp',
        'AllowTcpForwarding' => 'no',
        'X11Forwarding' => 'no'
      }
   }
  )
  cookbook        'sshd'
  source          'sshd_config.erb'
  action          :create
end
```

Here were the results:

```bash
             -UsePAM yes
             +SyslogFacility AUTH
       UseDNS no
             -GSSAPIAuthentication no
             +UsePAM yes
             +X11Forwarding yes
             +Match User fred
             +    X11Forwarding no
             +Match User john
             +    ChrootDirectory /srv
             +    ForceCommand internal-sftp
             +    AllowTcpForwarding no
             +    X11Forwarding no
             +
```

chr4 commented:
Nice! I can confirm that this is working with after adapting the resource according to the documentation. 

I'm looking forward to your contribution regarding `inspec`! If you plan to add tests, I'd delay the merge until the tests are finished, if you don't mind? I'll then release a new major version of the cookbook.

iamjohnnym commented:
Alright.  I've got a working version of `inspec` going.  I'll need to make some additional checks of content in the `sshd_config` based on os platform but that's topical.  I am a bit lost on how to implement other tests, such as validating using `sshd_config()` within the resource.  Any insight would be awesome.

Also, I stopped testing on a `Port` change because vagrant didnt like that, the inspec process was unable to log in to validate the tests.  Is there a better way here?

Here's my current PR in the works.

https://github.com/iamjohnnym/sshd/pull/1/files

chr4 commented:
Thanks! Let me know when you feel comfortable merging it. I've added some minor comments, hope they'll be helpful!

iamjohnnym commented:
Hey!  Thanks for taking a peek at that.  I made the changes you recommended and then proceeded to destroy that branch.  I had to recreate it and clean it up for that merge.  I believe I broke the `rspec` tests though and I'm not sure I know the best way to resolve them.  Any ideas there?  I was looking at using `delivery` too but I made some things fairly angry.

iamjohnnym commented:
Alright, I think the release is good to go.  I leave that decision to you.  If you decide changes need to be made, let me know.

chr4 commented:
Thanks! Looks good on first sight, I'll have a closer look next week :)

chr4 commented:
I did a major bump to version `v2.0.0` as this breaks existing configurations and released it!

Thanks so much, was a pleasure!
